### PR TITLE
feat: add feature deletion_protection in bigquery table

### PR DESCRIPTION
**Pull request by franviera92**

### CODE REVIEW

**Comment by comment-bot-dev** [comment]
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_documentation
 The documentation needs to be regenerated. Please run `make generate_docs`. 
 ```bash 
Checking for documentation generation
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' '--exclude=autogen' '--exclude=*.tfvars' /workspace/README.md /tmp/tmp.2gsq9qMmU7/generate_docs/workspace/README.md
168a169
> | deletion\_protection | Whether or not to allow Terraform to destroy the instance. Unless this field is set to false in Terraform state, a terraform destroy or terraform apply that would delete the instance will fail. | `bool` | `true` | no |
Error: Documentation generation has not been run, please run the
'make docker_generate_docs' command and commit the above changes.
 ``` 
- ⚠️check_terraform
 Failed Terraform check. More details below. 
 ```bash 
Running terraform fmt
Running terraform validate
terraform_validate . 
Success!
 The configuration is valid.
terraform_validate ./examples/basic_bq 
Success!
 The configuration is valid.
terraform_validate ./examples/basic_view 
Success!
 The configuration is valid.
terraform_validate ./examples/multiple_tables 
Error: 
Unsupported argument
  on ../../main.tf line 72, in resource "google_bigquery_table" "main":
  72:   
deletion_protection
 = var.deletion_protection
An argument named "deletion_protection" is not expected here.
terraform_validate ./modules/authorization 
Success!
 The configuration is valid.
terraform_validate ./modules/udf 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/full 
Error: 
Unsupported argument
  on ../../../main.tf line 72, in resource "google_bigquery_table" "main":
  72:   
deletion_protection
 = var.deletion_protection
An argument named "deletion_protection" is not expected here.
terraform_validate ./test/setup 
Success!
 The configuration is valid.
 ```

**Comment by franviera92** [comment]
Where are is failing?

**Comment by morgante** [comment]
@franviera92 See the comment ^^^

**Comment by franviera92** [comment]
> @franviera92 See the comment ^^^

Now which is fail?

**Comment by morgante** [comment]
```
       module.example.module.bigquery.google_bigquery_table.external_table["hive_example"]: Destroying... [id=projects/ci-bigquery-b500/datasets/foo/tables/hive_example]
       
       Error: cannot destroy instance without setting deletion_protection=false and running `terraform apply`
       
       
       
       Error: cannot destroy instance without setting deletion_protection=false and running `terraform apply`
       
       
       
       Error: cannot destroy instance without setting deletion_protection=false and running `terraform apply`
       
       
       
       Error: cannot destroy instance without setting deletion_protection=false and running `terraform apply`
       
       
       
       Error: cannot destroy instance without setting deletion_protection=false and running `terraform apply`
```

**Comment by franviera92** [comment]
It's ok :D

**Review by morgante — COMMENTED** [review]
Do we need per-resource variables? I think an overall variable would be sufficient.

**Comment by franviera92** [comment]
> Do we need per-resource variables? I think an overall variable would be sufficient.

Not necessarily, for example I need to activate protection for the tables and not the views that I create

**Comment by morgante** [comment]
> Not necessarily, for example I need to activate protection for the tables and not the views that I create

That makes sense, but perhaps it would make sense to still only have one variable but have it apply to the tables/dataset but not views?

It seems reasonable to not need data protection on views since deleting them doesn't cause data loss.

**Comment by franviera92** [comment]
> > Not necessarily, for example I need to activate protection for the tables and not the views that I create
> 
> That makes sense, but perhaps it would make sense to still only have one variable but have it apply to the tables/dataset but not views?
> 
> It seems reasonable to not need data protection on views since deleting them doesn't cause data loss.

Yes it is reasonable

**Comment by morgante** [comment]
> Yes it is reasonable

To be clear, I suggest we add an overall `deletion_protection` variable for datasets and tables. Then we don't set anything for views at all (no need for a separate variable).

**Comment by franviera92** [comment]
> > Yes it is reasonable
> 
> To be clear, I suggest we add an overall `deletion_protection` variable for datasets and tables. Then we don't set anything for views at all (no need for a separate variable).

I think it is better to separate the variables, delete_protection_table to tables and e_protection_view, I can create both in a single module but the restriction is different for each one

**Comment by morgante** [comment]
> I think it is better to separate the variables, delete_protection_table to tables and e_protection_view, I can create both in a single module but the restriction is different for each one

I don't think deletion protection is necessary for views at all.

**Comment by franviera92** [comment]
> > Creo que es mejor separar las variables, delete_protection_table a tablas y e_protection_view, puedo crear ambas en un solo módulo pero la restricción es diferente para cada una
> 
> No creo que la protección contra la eliminación sea necesaria para las vistas en absoluto.

we set hard deletion_protection = false for views? if so I like the idea

File: variables.tf [hunk 1]
```diff
@@ -43,6 +43,12 @@ variable "delete_contents_on_destroy" {
   default     = null
 }
 
+variable "deletion_protection" {
+  description = "(Whether or not to allow Terraform to destroy the instance. Unless this field is set to false in Terraform state, a terraform destroy or terraform apply that would delete the instance will fail"
```
Comment (morgante) [inline variables.tf:47]:
```suggestion
  description = "Whether or not to allow Terraform to destroy the instance. Unless this field is set to false in Terraform state, a terraform destroy or terraform apply that would delete the instance will fail"
```

File: README.md [hunk 2]
```diff
@@ -166,6 +166,7 @@ This module provisions a dataset and a list of tables with associated JSON schem
 | dataset\_name | Friendly name for the dataset being provisioned. | `string` | `null` | no |
 | default\_table\_expiration\_ms | TTL of tables using the dataset in MS | `number` | `null` | no |
 | delete\_contents\_on\_destroy | (Optional) If set to true, delete all the tables in the dataset when destroying the resource; otherwise, destroying the resource will fail if tables are present. | `bool` | `null` | no |
+| deletion\_protection | (Whether or not to allow Terraform to destroy the instance. Unless this field is set to false in Terraform state, a terraform destroy or terraform apply that would delete the instance will fail | `bool` | `false` | no |
```
Comment (morgante) [inline README.md:169]:
```suggestion
| deletion\_protection | Whether or not to allow Terraform to destroy the instance. Unless this field is set to false in Terraform state, a terraform destroy or terraform apply that would delete the instance will fail | `bool` | `false` | no |
```

**Review by morgante — APPROVED** [review]

**Comment by morgante** [comment]
> we set hard deletion_protection = false for views? if so I like the idea

Exactly, this looks good!

**Comment by m0ps** [comment]
Are there any plans to merge this PR and make a new release?

**Comment by franviera92** [comment]
I need this PR ☹️

**Review by olegeech-me — APPROVED** [review]

**Pull request merged**